File: src/services/bannerService.ts
/**
 * Banner 轮播图服务 - 处理 Banner 相关的 API 调用
 */

import { useAuthStore } from "../store/authStore";
import { config } from "../config/env";

const EXPO_PUBLIC_API_BASE_URL = config.EXPO_PUBLIC_API_BASE_URL;

// Banner 类型定义
export interface Banner {
  id: number;
  title: string;
  subtitle?: string;
  imageUrl: string;
  linkType: "NONE" | "POST" | "BRAND" | "EXTERNAL" | "SHOW";
  linkValue?: string;
  sortOrder: number;
  isActive: boolean;
  startTime?: string;
  endTime?: string;
  createdBy?: number;
  createdAt: string;
  updatedAt: string;
}

// 创建 Banner 请求类型
export interface CreateBannerRequest {
  title: string;
  subtitle?: string;
  image_url: string;
  link_type?: string;
  link_value?: string;
  sort_order?: number;
  is_active?: boolean;
  start_time?: string;
  end_time?: string;
}

// 更新 Banner 请求类型
export interface UpdateBannerRequest {
  title?: string;
  subtitle?: string;
  image_url?: string;
  link_type?: string;
  link_value?: string;
  sort_order?: number;
  is_active?: boolean;
  start_time?: string;
  end_time?: string;
}

// API 响应包装类型
interface ApiResponse<T> {
  code: number;
  message: string;
  data: T;
}

// 通用请求方法
async function request<T>(
  endpoint: string,
  options: RequestInit = {}
): Promise<T> {
  const url = `${EXPO_PUBLIC_API_BASE_URL}${endpoint}`;

  const headers: Record<string, string> = {
    "Content-Type": "application/json",
    Accept: "*/*",
    ...((options.headers as Record<string, string>) || {}),
  };

  // 自动添加 Authorization header（如果已登录）
  const token = useAuthStore.getState().getAccessToken();
  if (token) {
    headers["Authorization"] = `Bearer ${token}`;
  }

  const config: RequestInit = {
    ...options,
    headers,
  };

  try {
    const response = await fetch(url, config);
    const contentType = response.headers.get("content-type");

    if (!response.ok) {
      let errorMessage = "请求失败";

      if (contentType?.includes("application/json")) {
        const errorData = await response.json();
        errorMessage = errorData.message || errorData.error || errorMessage;
      } else {
        const text = await response.text();
        errorMessage = text || `HTTP ${response.status}`;
      }

      throw new Error(errorMessage);
    }

    if (contentType?.includes("application/json")) {
      const jsonResponse = await response.json();

      // 处理包装的 API 响应格式 { code, message, data }
      if (
        jsonResponse &&
        typeof jsonResponse === "object" &&
        "code" in jsonResponse
      ) {
        const apiResponse = jsonResponse as ApiResponse<T>;

        // 检查业务错误码
        if (apiResponse.code !== 0) {
          throw new Error(apiResponse.message || "请求失败");
        }

        // 返回 data 字段（如果存在）
        if ("data" in apiResponse) {
          return apiResponse.data;
        }
      }

      // 如果不是包装格式，直接返回
      return jsonResponse as T;
    }

    const text = await response.text();
    return text as unknown as T;
  } catch (error) {
    if (error instanceof Error) {
      throw error;
    }
    throw new Error("网络请求失败，请检查网络连接");
  }
}

// ==================== 公开接口 ====================

/**
 * 获取当前有效的 Banner 列表（前端展示用）
 * GET /api/banners/active
 */
export async function getActiveBanners(): Promise<Banner[]> {
  return request<Banner[]>("/api/banners/active", {
    method: "GET",
  });
}

// ==================== 管理员接口 ====================

/**
 * 获取所有 Banner 列表（管理员用）
 * GET /api/banners/admin/list
 */
export async function getAllBanners(): Promise<Banner[]> {
  return request<Banner[]>("/api/banners/admin/list", {
    method: "GET",
  });
}

/**
 * 获取单个 Banner 详情
 * GET /api/banners/admin/{bannerId}
 */
export async function getBanner(bannerId: number): Promise<Banner> {
  return request<Banner>(`/api/banners/admin/${bannerId}`, {
    method: "GET",
  });
}

/**
 * 创建 Banner
 * POST /api/banners/admin
 */
export async function createBanner(data: CreateBannerRequest): Promise<Banner> {
  return request<Banner>("/api/banners/admin", {
    method: "POST",
    body: JSON.stringify(data),
  });
}

/**
 * 更新 Banner
 * PUT /api/banners/admin/{bannerId}
 */
export async function updateBanner(
  bannerId: number,
  data: UpdateBannerRequest
): Promise<Banner> {
  return request<Banner>(`/api/banners/admin/${bannerId}`, {
    method: "PUT",
    body: JSON.stringify(data),
  });
}

/**
 * 删除 Banner
 * DELETE /api/banners/admin/{bannerId}
 */
export async function deleteBanner(bannerId: number): Promise<void> {
  return request<void>(`/api/banners/admin/${bannerId}`, {
    method: "DELETE",
  });
}

/**
 * 切换 Banner 启用状态
 * POST /api/banners/admin/{bannerId}/toggle
 */
export async function toggleBannerStatus(bannerId: number): Promise<Banner> {
  return request<Banner>(`/api/banners/admin/${bannerId}/toggle`, {
    method: "POST",
  });
}

/**
 * 重新排序 Banner
 * POST /api/banners/admin/reorder
 */
export async function reorderBanners(bannerIds: number[]): Promise<void> {
  return request<void>("/api/banners/admin/reorder", {
    method: "POST",
    body: JSON.stringify({ banner_ids: bannerIds }),
  });
}

// 导出 bannerService 对象
export const bannerService = {
  // 公开接口
  getActiveBanners,
  // 管理员接口
  getAllBanners,
  getBanner,
  createBanner,
  updateBanner,
  deleteBanner,
  toggleBannerStatus,
  reorderBanners,
};

export default bannerService;

File: backend/app/services/banner_service.py
"""
Banner 服务
"""
from typing import List, Optional
from datetime import datetime
from app.db.supabase import get_supabase
from app.schemas.banner import BannerCreate, BannerUpdate, BannerResponse


class BannerService:
    def __init__(self):
        self.db = get_supabase()

    def _format_banner(self, banner_data: dict) -> BannerResponse:
        """格式化 Banner 数据为前端格式"""
        return BannerResponse(
            id=banner_data["id"],
            title=banner_data["title"],
            subtitle=banner_data.get("subtitle"),
            imageUrl=banner_data["image_url"],
            linkType=banner_data.get("link_type", "NONE"),
            linkValue=banner_data.get("link_value"),
            sortOrder=banner_data.get("sort_order", 0),
            isActive=banner_data.get("is_active", True),
            startTime=banner_data.get("start_time"),
            endTime=banner_data.get("end_time"),
            createdBy=banner_data.get("created_by"),
            createdAt=banner_data["created_at"],
            updatedAt=banner_data["updated_at"]
        )

    def get_active_banners(self) -> List[BannerResponse]:
        """获取当前有效的 Banner 列表（前端展示用）"""
        now = datetime.utcnow().isoformat()
        
        # 查询启用且在有效时间范围内的 Banner
        result = self.db.table("banners") \
            .select("*") \
            .eq("is_active", True) \
            .order("sort_order", desc=False) \
            .execute()
        
        banners = []
        for b in result.data or []:
            # 检查时间范围
            start_time = b.get("start_time")
            end_time = b.get("end_time")
            
            # 如果设置了开始时间，检查是否已到开始时间
            if start_time and start_time > now:
                continue
            
            # 如果设置了结束时间，检查是否已过期
            if end_time and end_time < now:
                continue
            
            banners.append(self._format_banner(b))
        
        return banners

    def get_all_banners(self) -> List[BannerResponse]:
        """获取所有 Banner 列表（管理员用）"""
        result = self.db.table("banners") \
            .select("*") \
            .order("sort_order", desc=False) \
            .execute()
        
        return [self._format_banner(b) for b in result.data or []]

    def get_banner_by_id(self, banner_id: int) -> Optional[BannerResponse]:
        """根据 ID 获取 Banner"""
        result = self.db.table("banners") \
            .select("*") \
            .eq("id", banner_id) \
            .execute()
        
        if result.data:
            return self._format_banner(result.data[0])
        return None

    def create_banner(self, data: BannerCreate, created_by: int) -> BannerResponse:
        """创建 Banner"""
        insert_data = {
            "title": data.title,
            "subtitle": data.subtitle,
            "image_url": data.image_url,
            "link_type": data.link_type,
            "link_value": data.link_value,
            "sort_order": data.sort_order,
            "is_active": data.is_active,
            "start_time": data.start_time.isoformat() if data.start_time else None,
            "end_time": data.end_time.isoformat() if data.end_time else None,
            "created_by": created_by
        }
        
        result = self.db.table("banners").insert(insert_data).execute()
        
        if result.data:
            return self._format_banner(result.data[0])
        raise Exception("创建 Banner 失败")

    def update_banner(self, banner_id: int, data: BannerUpdate) -> Optional[BannerResponse]:
        """更新 Banner"""
        update_data = {}
        
        if data.title is not None:
            update_data["title"] = data.title
        if data.subtitle is not None:
            update_data["subtitle"] = data.subtitle
        if data.image_url is not None:
            update_data["image_url"] = data.image_url
        if data.link_type is not None:
            update_data["link_type"] = data.link_type
        if data.link_value is not None:
            update_data["link_value"] = data.link_value
        if data.sort_order is not None:
            update_data["sort_order"] = data.sort_order
        if data.is_active is not None:
            update_data["is_active"] = data.is_active
        if data.start_time is not None:
            update_data["start_time"] = data.start_time.isoformat()
        if data.end_time is not None:
            update_data["end_time"] = data.end_time.isoformat()
        
        if not update_data:
            return self.get_banner_by_id(banner_id)
        
        result = self.db.table("banners") \
            .update(update_data) \
            .eq("id", banner_id) \
            .execute()
        
        if result.data:
            return self._format_banner(result.data[0])
        return None

    def delete_banner(self, banner_id: int) -> bool:
        """删除 Banner"""
        result = self.db.table("banners") \
            .delete() \
            .eq("id", banner_id) \
            .execute()
        
        return bool(result.data)

    def toggle_banner_status(self, banner_id: int) -> Optional[BannerResponse]:
        """切换 Banner 启用状态"""
        # 先获取当前状态
        banner = self.get_banner_by_id(banner_id)
        if not banner:
            return None
        
        # 切换状态
        new_status = not banner.isActive
        result = self.db.table("banners") \
            .update({"is_active": new_status}) \
            .eq("id", banner_id) \
            .execute()
        
        if result.data:
            return self._format_banner(result.data[0])
        return None

    def reorder_banners(self, banner_ids: List[int]) -> bool:
        """重新排序 Banner"""
        try:
            for index, banner_id in enumerate(banner_ids):
                self.db.table("banners") \
                    .update({"sort_order": index}) \
                    .eq("id", banner_id) \
                    .execute()
            return True
        except Exception:
            return False


# 单例
banner_service = BannerService()
